File: OTEAServer/Controllers/IndicatorsEvaluationRegsController.cs
using Microsoft.AspNetCore.Mvc;
using OTEAServer.Misc;
using OTEAServer.Models;
namespace OTEAServer.Controllers
{

    /// <summary>
    /// Controller for indicators evaluation registers operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>
    [ApiController]
    [Route("IndicatorsEvaluationsRegs")]
    public class IndicatorsEvaluationsRegsController : ControllerBase
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;

        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context">Database context</param>
        public IndicatorsEvaluationsRegsController(DatabaseContext context)
        {
            _context = context;
        }

        /// <summary>
        /// Method that obtains all the registers
        /// </summary>
        /// <returns>Registrations list</returns>
        [HttpGet("all")]
        public IActionResult GetAll() {
            try
            {
                var IndicatorsEvaluationsRegs = _context.IndicatorsEvaluationsRegs.ToList();
                return Ok(IndicatorsEvaluationsRegs);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains all registers of an indicators evaluation
        /// </summary>
        /// <param name="evaluationDate">Evaluation date in timestamp</param>
        /// <param name="idEvaluatorTeam">Evaluator team identifier</param>
        /// <param name="idEvaluatorOrganization">Identifier of the evaluator organization that will do the indicators evaluation</param>
        /// <param name="orgTypeEvaluator">Organization type of the evaluator organization that will do the indicators evaluation.It will be always "EVALUATOR"</param>
        /// <param name="idEvaluatedOrganization">Identifier of the external organization that will recieve the indicators evaluation</param>
        /// <param name="orgTypeEvaluated">Organization type of the external organization that will recieve the indicators evaluation. It will be always "EVALUATED"</param>
        /// <param name="illness">Illness of the external organization that will recieve the indicators evaluation. In case of Fundación Miradas, it will be "AUTISM"</param>
        /// <param name="idCenter">Center identifier of the external organization</param>
        /// <returns>Indicators list</returns>
        [HttpGet("indEval")]
        public IActionResult GetAllByIndicatorsEvaluation([FromQuery] long evaluationDate, [FromQuery] int idEvaluatorTeam, [FromQuery] int idEvaluatorOrganization, [FromQuery] string orgTypeEvaluator, [FromQuery] int idEvaluatedOrganization, [FromQuery] string orgTypeEvaluated, [FromQuery] string illness, [FromQuery] int idCenter) {
            try
            {
                var IndicatorsEvaluationsRegs = _context.IndicatorsEvaluationsRegs.Where(r => r.evaluationDate == evaluationDate && r.idEvaluatorTeam == idEvaluatorTeam && r.idEvaluatorOrganization == idEvaluatorOrganization && r.orgTypeEvaluator == orgTypeEvaluator && r.idEvaluatedOrganization == idEvaluatedOrganization && r.orgTypeEvaluated == orgTypeEvaluated && r.illness == illness && r.idCenter == idCenter).ToList();
                return Ok(IndicatorsEvaluationsRegs);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }


        /// <summary>
        /// Gets an indicators evaluation register
        /// </summary>
        /// <param name="evaluationDate">Evaluation date in timestamp</param>
        /// <param name="idEvaluatorTeam">Evaluator team identifier</param>
        /// <param name="idEvaluatorOrganization">Identifier of the evaluator organization that will do the indicators evaluation</param>
        /// <param name="orgTypeEvaluator">Organization type of the evaluator organization that will do the indicators evaluation. It will be always "EVALUATOR"</param>
        /// <param name="idEvaluatedOrganization"></param>
        /// <param name="orgTypeEvaluated">Identifier of the external organization that will recieve the indicators evaluation</param>
        /// <param name="illness">Illness of the external organization that will recieve the indicators evaluation. In case of Fundación Miradas, it will be "AUTISM"</param>
        /// <param name="idCenter">Center identifier of the external organization</param>
        /// <param name="idSubSubAmbit">Second level division of the ambit. It will be -1 if there is no division</param>
        /// <param name="idSubAmbit">First level division of the ambit. It will be -1 if there is no division</param>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <param name="idIndicator">Indicator identifier</param>
        /// <param name="idEvidence">Evidence identifier</param>
        /// <param name="indicatorVersion">Indicator version</param>
        /// <returns>Register if success, null if not</returns>

        [HttpGet("get")]
        public ActionResult<IndicatorsEvaluationReg> Get([FromQuery] long evaluationDate, [FromQuery] int idEvaluatorTeam, [FromQuery] int idEvaluatorOrganization, [FromQuery] string orgTypeEvaluator, [FromQuery] int idEvaluatedOrganization, [FromQuery] string orgTypeEvaluated, [FromQuery] string illness, [FromQuery] int idCenter, [FromQuery] int idSubSubAmbit, [FromQuery] int idSubAmbit, [FromQuery] int idAmbit, [FromQuery] int idIndicator, [FromQuery] int idEvidence, [FromQuery] int indicatorVersion) {
            try
            {
                var indicatorsEvaluation = _context.IndicatorsEvaluationsRegs.FirstOrDefault(r => r.evaluationDate == evaluationDate && r.idEvaluatorTeam == idEvaluatorTeam && r.idEvaluatorOrganization == idEvaluatorOrganization && r.orgTypeEvaluator == orgTypeEvaluator && r.idEvaluatedOrganization == idEvaluatedOrganization && r.orgTypeEvaluated == orgTypeEvaluated && r.illness == illness && r.idCenter == idCenter && r.idSubSubAmbit == idSubSubAmbit && r.idSubAmbit == idSubAmbit && r.idAmbit == idAmbit && r.idIndicator == idIndicator && r.idEvidence == idEvidence && r.indicatorVersion == indicatorVersion);

                if (indicatorsEvaluation == null)
                    return NotFound();

                return indicatorsEvaluation;
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that appends a new indicators evaluation register
        /// </summary>
        /// <param name="indicatorsEvaluationReg">Indicators evaluation register</param>
        /// <returns>Register if success, null if not</returns>
        [HttpPost]
        public ActionResult<IndicatorsEvaluationReg> Create([FromBody] IndicatorsEvaluationReg indicatorsEvaluationReg) {
            try
            {
                _context.IndicatorsEvaluationsRegs.Add(indicatorsEvaluationReg);
                _context.SaveChanges();
                return CreatedAtAction(nameof(Get), new
                {
                    evaluationDate = indicatorsEvaluationReg.evaluationDate,
                    idEvaluatorTeam = indicatorsEvaluationReg.idEvaluatorTeam,
                    idEvaluatorOrganization = indicatorsEvaluationReg.idEvaluatorOrganization,
                    orgTypeEvaluator = indicatorsEvaluationReg.orgTypeEvaluator,
                    idEvaluatedOrganization = indicatorsEvaluationReg.idEvaluatedOrganization,
                    orgTypeEvaluated = indicatorsEvaluationReg.orgTypeEvaluated,
                    illness = indicatorsEvaluationReg.illness,
                    idCenter = indicatorsEvaluationReg.idCenter,
                    idSubSubAmbit = indicatorsEvaluationReg.idSubSubAmbit,
                    idSubAmbit = indicatorsEvaluationReg.idSubAmbit,
                    idAmbit = indicatorsEvaluationReg.idAmbit,
                    idIndicator = indicatorsEvaluationReg.idIndicator,
                    idEvidence = indicatorsEvaluationReg.idEvidence,
                    indicatorVersion = indicatorsEvaluationReg.indicatorVersion
                }, indicatorsEvaluationReg);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that appends new indicators evaluation register
        /// </summary>
        /// <param name="regs">Indicators evaluation registers</param>
        /// <returns>Register if success, null if not</returns>
        [HttpPost("fromList")]
        public IActionResult CreateRegs([FromBody] List<IndicatorsEvaluationReg> regs)
        {
            try
            {
                foreach (IndicatorsEvaluationReg reg in regs)
                {
                    _context.IndicatorsEvaluationsRegs.Add(reg);
                }
                _context.SaveChanges();
                return Ok(regs);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }



        /// <summary>
        /// 
        /// </summary>
        /// <param name="evaluationDate">Evaluation date in timestamp</param>
        /// <param name="idEvaluatorTeam">Evaluator team identifier</param>
        /// <param name="idEvaluatorOrganization">Identifier of the evaluator organization that will do the indicators evaluation</param>
        /// <param name="orgTypeEvaluator">Organization type of the evaluator organization that will do the indicators evaluation. It will be always "EVALUATOR"</param>
        /// <param name="idEvaluatedOrganization"></param>
        /// <param name="orgTypeEvaluated">Identifier of the external organization that will recieve the indicators evaluation</param>
        /// <param name="illness">Illness of the external organization that will recieve the indicators evaluation. In case of Fundación Miradas, it will be "AUTISM"</param>
        /// <param name="idCenter">Center identifier of the external organization</param>
        /// <param name="idSubSubAmbit">Second level division of the ambit. It will be -1 if there is no division</param>
        /// <param name="idSubAmbit">First level division of the ambit. It will be -1 if there is no division</param>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <param name="idIndicator">Indicator identifier</param>
        /// <param name="idEvidence">Evidence identifier</param>
        /// <param name="indicatorVersion">Indicator version</param>
        /// <param name="indicatorsEvaluationReg">Indicators evaluation registration</param>
        /// <returns>Register if success, null if not</returns>
        [HttpPut]
        public ActionResult<IndicatorsEvaluationReg> Update([FromQuery] long evaluationDate, [FromQuery] int idEvaluatorTeam, [FromQuery] int idEvaluatorOrganization, [FromQuery] string orgTypeEvaluator, [FromQuery] int idEvaluatedOrganization, [FromQuery] string orgTypeEvaluated, [FromQuery] string illness, [FromQuery] int idCenter, [FromQuery] int idSubSubAmbit, [FromQuery] int idSubAmbit, [FromQuery] int idAmbit, [FromQuery] int idIndicator, [FromQuery] int idEvidence, [FromQuery] int indicatorVersion, [FromBody] IndicatorsEvaluationReg indicatorsEvaluationReg) {
            try
            {
                var existingIndicatorsEvaluationReg = _context.IndicatorsEvaluationsRegs.FirstOrDefault(r => r.evaluationDate == evaluationDate && r.idEvaluatorTeam == idEvaluatorTeam && r.idEvaluatorOrganization == idEvaluatorOrganization && r.orgTypeEvaluator == orgTypeEvaluator && r.idEvaluatedOrganization == idEvaluatedOrganization && r.orgTypeEvaluated == orgTypeEvaluated && r.illness == illness && r.idCenter == idCenter && r.idSubSubAmbit == idSubSubAmbit && r.idSubAmbit == idSubAmbit && r.idAmbit == idAmbit && r.idIndicator == idIndicator && r.idEvidence == idEvidence && r.indicatorVersion == indicatorVersion);
                if (existingIndicatorsEvaluationReg is null)
                    return NotFound();

                existingIndicatorsEvaluationReg.evaluationDate = indicatorsEvaluationReg.evaluationDate;
                existingIndicatorsEvaluationReg.idEvaluatorTeam = indicatorsEvaluationReg.idEvaluatorTeam;
                existingIndicatorsEvaluationReg.idEvaluatorOrganization = indicatorsEvaluationReg.idEvaluatorOrganization;
                existingIndicatorsEvaluationReg.orgTypeEvaluator = indicatorsEvaluationReg.orgTypeEvaluator;
                existingIndicatorsEvaluationReg.idEvaluatedOrganization = indicatorsEvaluationReg.idEvaluatedOrganization;
                existingIndicatorsEvaluationReg.orgTypeEvaluated = indicatorsEvaluationReg.orgTypeEvaluated;
                existingIndicatorsEvaluationReg.illness = indicatorsEvaluationReg.illness;
                existingIndicatorsEvaluationReg.idCenter = indicatorsEvaluationReg.idCenter;
                existingIndicatorsEvaluationReg.idSubSubAmbit = indicatorsEvaluationReg.idSubSubAmbit;
                existingIndicatorsEvaluationReg.idSubAmbit = indicatorsEvaluationReg.idSubAmbit;
                existingIndicatorsEvaluationReg.idAmbit = indicatorsEvaluationReg.idAmbit;
                existingIndicatorsEvaluationReg.idIndicator = indicatorsEvaluationReg.idIndicator;
                existingIndicatorsEvaluationReg.idEvidence = indicatorsEvaluationReg.idEvidence;
                existingIndicatorsEvaluationReg.isMarked = indicatorsEvaluationReg.isMarked;
                existingIndicatorsEvaluationReg.indicatorVersion = indicatorsEvaluationReg.indicatorVersion;
                _context.SaveChanges();
                return Ok(existingIndicatorsEvaluationReg);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }


        /// <summary>
        /// Method that deletes an indicators evaluation register
        /// </summary>
        /// <param name="evaluationDate">Evaluation date in timestamp</param>
        /// <param name="idEvaluatorTeam">Evaluator team identifier</param>
        /// <param name="idEvaluatorOrganization">Identifier of the evaluator organization that will do the indicators evaluation</param>
        /// <param name="orgTypeEvaluator">Organization type of the evaluator organization that will do the indicators evaluation. It will be always "EVALUATOR"</param>
        /// <param name="idEvaluatedOrganization"></param>
        /// <param name="orgTypeEvaluated">Identifier of the external organization that will recieve the indicators evaluation</param>
        /// <param name="illness">Illness of the external organization that will recieve the indicators evaluation. In case of Fundación Miradas, it will be "AUTISM"</param>
        /// <param name="idCenter">Center identifier of the external organization</param>
        /// <param name="idSubSubAmbit">Second level division of the ambit. It will be -1 if there is no division</param>
        /// <param name="idSubAmbit">First level division of the ambit. It will be -1 if there is no division</param>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <param name="idIndicator">Indicator identifier</param>
        /// <param name="idEvidence">Evidence identifier</param>
        /// <param name="indicatorVersion">Indicator version</param>
        /// <returns>Register if success, null if not</returns>
        [HttpDelete]
        public ActionResult<IndicatorsEvaluationReg> Delete([FromQuery] long evaluationDate, [FromQuery] int idEvaluatorTeam, [FromQuery] int idEvaluatorOrganization, [FromQuery] string orgTypeEvaluator, [FromQuery] int idEvaluatedOrganization, [FromQuery] string orgTypeEvaluated, [FromQuery] string illness, [FromQuery] int idCenter, [FromQuery] int idSubSubAmbit, [FromQuery] int idSubAmbit, [FromQuery] int idAmbit, [FromQuery] int idIndicator, [FromQuery] int idEvidence, [FromQuery] int indicatorVersion) {
            try
            {
                var indicatorsEvaluation = _context.IndicatorsEvaluationsRegs.FirstOrDefault(r => r.evaluationDate == evaluationDate && r.idEvaluatorTeam == idEvaluatorTeam && r.idEvaluatorOrganization == idEvaluatorOrganization && r.orgTypeEvaluator == orgTypeEvaluator && r.idEvaluatedOrganization == idEvaluatedOrganization && r.orgTypeEvaluated == orgTypeEvaluated && r.illness == illness && r.idCenter == idCenter && r.idSubSubAmbit == idSubSubAmbit && r.idSubAmbit == idSubAmbit && r.idAmbit == idAmbit && r.idIndicator == idIndicator && r.idEvidence == idEvidence && r.indicatorVersion == indicatorVersion);

                if (indicatorsEvaluation is null)
                    return NotFound();

                _context.IndicatorsEvaluationsRegs.Remove(indicatorsEvaluation);
                _context.SaveChanges();
                return NoContent();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

    }
}

File: connection/src/main/java/otea/connection/controller/IndicatorsEvaluationRegsController.java
package otea.connection.controller;


import java.io.IOException;
;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.indicators.Indicator;
import cli.indicators.IndicatorsEvaluationReg;
import otea.connection.ConnectionClient;
import otea.connection.api.IndicatorsEvaluationRegsApi;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.http.Query;

/**
 * Controller for indicators evaluation registers
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 * */
public class IndicatorsEvaluationRegsController {

    /**Indicators evaluation registers api to connect to the server*/
    private static IndicatorsEvaluationRegsApi api;

    /**Controller instance*/
    private static IndicatorsEvaluationRegsController instance;

    /**Class constructor*/
    private IndicatorsEvaluationRegsController(){
        api= ConnectionClient.getInstance().getRetrofit().create(IndicatorsEvaluationRegsApi.class);
    }

    /**
     * Method that obtains the singleton instance of the controller
     *
     * @return Controller instance
     * */
    public static IndicatorsEvaluationRegsController getInstance(){
        if(instance==null){
            synchronized (IndicatorsEvaluationRegsController.class){
                if(instance==null){
                    instance=new IndicatorsEvaluationRegsController();
                }
            }
        }
        return instance;
    }

    /**Refresh API*/
    public static void refreshApi(){
        instance=new IndicatorsEvaluationRegsController();
    }
    
    /**
     * Method that obtains all the registers
     *
     * @return Registrations list
     * */
    public static List<IndicatorsEvaluationReg> GetAll(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<IndicatorsEvaluationReg>> callable = new Callable<List<IndicatorsEvaluationReg>>() {
            @Override
            public List<IndicatorsEvaluationReg> call() throws Exception {
                Call<List<IndicatorsEvaluationReg>> call=api.GetAll();
                Response<List<IndicatorsEvaluationReg>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<IndicatorsEvaluationReg>> future = executor.submit(callable);
            List<IndicatorsEvaluationReg> list = future.get();
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that obtains all registers of an indicators evaluation
     *
     * @param evaluationDate - Evaluation date in timestamp
     * @param idEvaluatorTeam - Evaluator team identifier
     * @param idEvaluatedOrganization - Identifier of the evaluator organization that will do the indicators evaluation
     * @param orgTypeEvaluator - Organization type of the evaluator organization that will do the indicators evaluation. It will be always "EVALUATOR"
     * @param idEvaluatorOrganization - Identifier of the external organization that will recieve the indicators evaluation
     * @param orgTypeEvaluated - Organization type of the external organization that will recieve the indicators evaluation. It will be always "EVALUATED"
     * @param illness - Illness of the external organization that will recieve the indicators evaluation. In case of Fundación Miradas, it will be "AUTISM"
     * @param idCenter - Center identifier of the external organization
     * @return Indicators list
     * */
    public static List<IndicatorsEvaluationReg> GetAllByIndicatorsEvaluation(long evaluationDate, int idEvaluatorTeam, int idEvaluatorOrganization, String orgTypeEvaluator, int idEvaluatedOrganization, String orgTypeEvaluated, String illness, int idCenter){

        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<IndicatorsEvaluationReg>> callable = new Callable<List<IndicatorsEvaluationReg>>() {
            @Override
            public List<IndicatorsEvaluationReg> call() throws Exception {
                Call<List<IndicatorsEvaluationReg>> call=api.GetAllByIndicatorsEvaluation(evaluationDate, idEvaluatorTeam, idEvaluatorOrganization, orgTypeEvaluator, idEvaluatedOrganization, orgTypeEvaluated, illness, idCenter);
                Response<List<IndicatorsEvaluationReg>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<IndicatorsEvaluationReg>> future = executor.submit(callable);
            List<IndicatorsEvaluationReg> list = future.get();
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Gets an indicators evaluation register
     *
     * @param evaluationDate - Evaluation date in timestamp
     * @param idEvaluatorTeam - Evaluator team identifier
     * @param idEvaluatedOrganization - Identifier of the evaluator organization that will do the indicators evaluation
     * @param orgTypeEvaluator - Organization type of the evaluator organization that will do the indicators evaluation. It will be always "EVALUATOR"
     * @param idEvaluatorOrganization - Identifier of the external organization that will recieve the indicators evaluation
     * @param orgTypeEvaluated - Organization type of the external organization that will recieve the indicators evaluation. It will be always "EVALUATED"
     * @param illness - Illness of the external organization that will recieve the indicators evaluation. In case of Fundación Miradas, it will be "AUTISM"
     * @param idCenter - Center identifier of the external organization
     * @param idSubSubAmbit - Second level division of the ambit. It will be -1 if there is no division
     * @param idSubAmbit - First level division of the ambit. It will be -1 if there is no division
     * @param idAmbit - Ambit identifier
     * @param idIndicator - Indicator identifier
     * @param idEvidence - Evidence identifier
     * @param indicatorVersion - Indicator version
     * @return Register if success, null if not
     * */
    public static IndicatorsEvaluationReg Get(long evaluationDate, int idEvaluatorTeam, int idEvaluatorOrganization, String orgTypeEvaluator, int idEvaluatedOrganization, String orgTypeEvaluated, String illness, int idCenter, int idSubSubAmbit, int idSubAmbit,int idAmbit, int idIndicator, int idEvidence, int indicatorVersion){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<IndicatorsEvaluationReg> callable = new Callable<IndicatorsEvaluationReg>() {
            @Override
            public IndicatorsEvaluationReg call() throws Exception {
                Call<IndicatorsEvaluationReg> call = api.Get(evaluationDate, idEvaluatorTeam, idEvaluatorOrganization, orgTypeEvaluator, idEvaluatedOrganization, orgTypeEvaluated, illness, idCenter, idSubSubAmbit, idSubAmbit, idAmbit, idIndicator, idEvidence, indicatorVersion);
                Response<IndicatorsEvaluationReg> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<IndicatorsEvaluationReg> future = executor.submit(callable);
            IndicatorsEvaluationReg result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that appends a new indicators evaluation register
     *
     * @param indicatorsEvaluationReg - Indicators evaluation register
     * @return Register if success, null if not
     * */
    public static IndicatorsEvaluationReg Create(IndicatorsEvaluationReg indicatorsEvaluationReg){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<IndicatorsEvaluationReg> callable = new Callable<IndicatorsEvaluationReg>() {
            @Override
            public IndicatorsEvaluationReg call() throws Exception {
                Call<IndicatorsEvaluationReg> call = api.Create(indicatorsEvaluationReg);
                Response<IndicatorsEvaluationReg> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<IndicatorsEvaluationReg> future = executor.submit(callable);
            IndicatorsEvaluationReg result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Updates an indicators evaluation register
     *
     * @param evaluationDate - Evaluation date in timestamp
     * @param idEvaluatorTeam - Evaluator team identifier
     * @param idEvaluatedOrganization - Identifier of the evaluator organization that will do the indicators evaluation
     * @param orgTypeEvaluator - Organization type of the evaluator organization that will do the indicators evaluation. It will be always "EVALUATOR"
     * @param idEvaluatorOrganization - Identifier of the external organization that will recieve the indicators evaluation
     * @param orgTypeEvaluated - Organization type of the external organization that will recieve the indicators evaluation. It will be always "EVALUATED"
     * @param illness - Illness of the external organization that will recieve the indicators evaluation. In case of Fundación Miradas, it will be "AUTISM"
     * @param idCenter - Center identifier of the external organization
     * @param idSubSubAmbit - Second level division of the ambit. It will be -1 if there is no division
     * @param idSubAmbit - First level division of the ambit. It will be -1 if there is no division
     * @param idAmbit - Ambit identifier
     * @param idIndicator - Indicator identifier
     * @param idEvidence - Evidence identifier
     * @param indicatorVersion - Indicator version
     * @param indicatorsEvaluationReg - Indicators evaluation registration
     * @return Register if success, null if not
     * */
    public static IndicatorsEvaluationReg Update(long evaluationDate, int idEvaluatorTeam, int idEvaluatorOrganization, String orgTypeEvaluator, int idEvaluatedOrganization, String orgTypeEvaluated, String illness, int idCenter, int idSubSubAmbit, int idSubAmbit, int idAmbit, int idIndicator, int idEvidence, int indicatorVersion, IndicatorsEvaluationReg indicatorsEvaluationReg){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<IndicatorsEvaluationReg> callable = new Callable<IndicatorsEvaluationReg>() {
            @Override
            public IndicatorsEvaluationReg call() throws Exception {
                Call<IndicatorsEvaluationReg> call = api.Update(evaluationDate, idEvaluatorTeam, idEvaluatorOrganization, orgTypeEvaluator, idEvaluatedOrganization, orgTypeEvaluated, illness, idCenter, idSubSubAmbit, idSubAmbit, idAmbit, idIndicator, idEvidence, indicatorVersion, indicatorsEvaluationReg);
                Response<IndicatorsEvaluationReg> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<IndicatorsEvaluationReg> future = executor.submit(callable);
            IndicatorsEvaluationReg result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }


    /**
     * Method that deletes an indicators evaluation register
     *
     * @param evaluationDate - Evaluation date in timestamp
     * @param idEvaluatorTeam - Evaluator team identifier
     * @param idEvaluatedOrganization - Identifier of the evaluator organization that will do the indicators evaluation
     * @param orgTypeEvaluator - Organization type of the evaluator organization that will do the indicators evaluation. It will be always "EVALUATOR"
     * @param idEvaluatorOrganization - Identifier of the external organization that will recieve the indicators evaluation
     * @param orgTypeEvaluated - Organization type of the external organization that will recieve the indicators evaluation. It will be always "EVALUATED"
     * @param illness - Illness of the external organization that will recieve the indicators evaluation. In case of Fundación Miradas, it will be "AUTISM"
     * @param idCenter - Center identifier of the external organization
     * @param idSubSubAmbit - Second level division of the ambit. It will be -1 if there is no division
     * @param idSubAmbit - First level division of the ambit. It will be -1 if there is no division
     * @param idAmbit - Ambit identifier
     * @param idIndicator - Indicator identifier
     * @param idEvidence - Evidence identifier
     * @param indicatorVersion - Indicator version
     * @return Register if success, null if not
     * */
    public static IndicatorsEvaluationReg Delete(long evaluationDate, int idEvaluatorTeam, int idEvaluatorOrganization, String orgTypeEvaluator, int idEvaluatedOrganization, String orgTypeEvaluated, String illness, int idCenter, int idSubSubAmbit, int idSubAmbit, int idAmbit, int idIndicator, int idEvidence, int indicatorVersion){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<IndicatorsEvaluationReg> callable = new Callable<IndicatorsEvaluationReg>() {
            @Override
            public IndicatorsEvaluationReg call() throws Exception {
                Call<IndicatorsEvaluationReg> call = api.Delete(evaluationDate, idEvaluatorTeam, idEvaluatorOrganization, orgTypeEvaluator, idEvaluatedOrganization, orgTypeEvaluated, illness, idCenter, idSubSubAmbit, idSubAmbit, idAmbit, idIndicator, idEvidence, indicatorVersion);
                Response<IndicatorsEvaluationReg> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<IndicatorsEvaluationReg> future = executor.submit(callable);
            IndicatorsEvaluationReg result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

}
